### PR TITLE
gh-71494: string.Formatter unnumbered key/attributes

### PR DESCRIPTION
**Pull request by qm2k**
https://bugs.python.org/issue27307

* Issue: gh-71494

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@qm2k

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by qm2k** [comment]
Blurb is trivial, but just in case the following test can be used:
> Add support for format strings like '{.name}' and '{[1]}' already supported by str.format to string.Formatter

**Comment by qm2k** [comment]
Rationale: we need to preserve the reenterability (which I think is important, e.g. in the looging use cases), and also the existing user subclasses that expect `''` to be converted to `'0'` (etc.) before the `get_field(...)` call. Cases where `field_name` starts with open bracket or dot would still change behaviour, but there's nothing we can do about it since original base class implementation raised exception on them and that is exactly what we are trying to change.

The alternative would be to add the missing reference in the `get_field(...)` itself, but it's hard to do since this function doesn't receive the numbering source, and adding new argument with it would break existing user subclasses. Passing the numbering source in a class attribute would break reenterability. It is possible to use some thread-local/traceback-keyed numbering sources stack attribute to preserve reenterability and (some) thread-safety, but it would hugely counterintuitive and likely quite fragile. Besides, changing the default `get_field(...)` implementation would not *fix* existing user subclasses that redefine this function. Therefore the only way to go was to fix the `field_name` before the `get_field(...)` call, which is what I did.

**Review by abrahammurciano — APPROVED** [review]

File: Lib/string.py [hunk 1]
```diff
@@ -186,19 +186,41 @@ def get_identifiers(self):
...
+        result = self._vformat(format_string, args, kwargs, used_args, 2)
         self.check_unused_args(used_args, args, kwargs)
         return result
 
     def _vformat(self, format_string, args, kwargs, used_args, recursion_depth,
-                 auto_arg_index=0):
+                 auto_number=None):
         if recursion_depth < 0:
             raise ValueError('Max string recursion exceeded')
+        if not auto_number:
```
Comment (dg-pb) [inline Lib/string.py]:
```suggestion
        if auto_number is None:
```

File: Lib/string.py [hunk 2]
```diff
@@ -186,19 +186,41 @@ def get_identifiers(self):
 # The field name parser is implemented in _string.formatter_field_name_split
 
 class Formatter:
+
+    class AutoNumber:
```
Comment (dg-pb) [inline Lib/string.py]: Maybe a private class at module level named `_AutoIndex`?

File: Lib/string.py [hunk 3]
```diff
@@ -186,19 +186,41 @@ def get_identifiers(self):
...
+        def __init__(self):
```
Comment (dg-pb) [inline Lib/string.py]:
```suggestion
        def __init__(self, index):
            self.index = index
```

File: Lib/string.py [hunk 4]
```diff
@@ -186,19 +186,41 @@ def get_identifiers(self):
...
+            self.field_number = 0
+
+        def __next__(self):
+            if self.field_number is False:
+                raise ValueError('cannot switch from manual field '
+                                 'specification to automatic field '
+                                 'numbering')
+            self.field_number += 1
```
Comment (dg-pb) [inline Lib/string.py]:
```suggestion
            try:
                return index
            finally:
                self.index += 1
```

File: Lib/string.py (hunk 1 above)
Comment (dg-pb) [reply Lib/string.py]: You might want to leave same arg name, which indicates starting index. I am not sure whether it is used, but this breaks backwards compatibility (for a private method, but still safer not to mess with it).

```python
def vformat(self, ..., auto_arg_index=0):
    ...
    if type(auto_arg_index) is int:
        auto_arg_index = _AutoIndex(auto_arg_index)
    ...
```

File: Lib/string.py [hunk 5]
```diff
@@ -288,6 +296,23 @@ def parse(self, format_string):
...
 
 
+    # given a field_name and auto_number, return a version
+    # starting with a name or index, taken from auto_number
+    # if necessary; calls auto_number.set_manual()
+    # if field_name already contains index
+    #  field_name:   the field being checked, e.g. "",
+    #                ".name" or "[3]"
+    #  auto_number   numbering source
+    def ensure_reference(self, field_name, auto_number):
```
Comment (dg-pb) [inline Lib/string.py]: It was quite neat and clear the way it was (part of the loop). Given some stuff is in the separate `Indexer` now, it is even clearer. Thus, I don't think refactoring into separate method is needed. Also, the name `ensure_reference` is more confusing than helpful to me.

Also, not sure if extra comments are needed here. The logic is straight forward.

File: Lib/string.py (hunk 2 above)
Comment (dg-pb) [reply Lib/string.py]: Or single-item-list could even be better - less changes and simpler.

File: Lib/string.py (hunk 5 above)
Comment (qm2k) [reply Lib/string.py]: Hi! Thanks a lot for reviewing this. Both me and Python have moved quite a lot since this PR was created, and I may need some time to swap in the context and comment on details. However, just wanted to let you know that your effort looking into this is much appreciated.

Comment (dg-pb) [reply Lib/string.py]: No worries, I found 3 open duplicate issues (after just opening one myself) and started looking at what has been done.

It doesn't seem that this particular part had any developments since.
Everything that you wrote seems to be still valid.

Are you still keen on following this through?

Comment (qm2k) [reply Lib/string.py]: I will try to find some time to look at it this weekend, if not — feel free to take it. 

Yes, this is quite a niche feature, but when you need it — you need it.

**Comment by dg-pb** [comment]
I see now that you have copied `AutoNumber` concept from `C`.

Personally, I like your changes, but for this to go through it would probably be best to only leave what is necessary. Which to me seems to be limited to `auto_reference` body, which can be placed in the main loop.

**Comment by qm2k** [comment]
> Personally, I like your changes, but for this to go through it would probably be best to only leave what is necessary. Which to me seems to be limited to `auto_reference` body, which can be placed in the main loop.

I am starting to remember my motivation for separate class and method: it was to provide access to corresponding parts to the user for further customization. I think nobody uses `string.Formatter` as it is, since by itself it is no better than `str.format`; most users probably subclass it and override some methods. I cannot however remember if my specific use case required further customization of automatic numbering or not.

I went ahead and made the change absolutely minimal which should address all of the issues raised in your comments. If somebody needs additional entry points for customization they are always welcome to submit a pull request :))) Please tell if you like the new version better.

**Comment by dg-pb** [comment]
> I am starting to remember my motivation for separate class and method

That was my starting point as well.
I wanted to pass `auto_arg_index` to `get_field`.
Issue is that it is very hard to do something without breaking backwards compatibility.

I have a use case in mind for which these changes don't cut it.
Namely, `PartialFormatter`, which needs `get_field` to have both original `field_name` and also `auto_arg_index` adjustments.

Then I thought of something along the lines of `AutoNumber` to pass it easily to methods, so initially it made sense, but eventually realised that this path is a bit tricky.

Long story short, I will spend some more time, maybe there is some good path forward. If there is not, simply pushing minimal bug fix is probably better than nothing at all.

**Comment by dg-pb** [comment]
> Please tell if you like the new version better.

I think this is great.
This will fix the issue at hand with minimal changes.
And yes, agreed, other things are best left to be done in separate PRs.

File: Lib/string.py [hunk 6]
```diff
@@ -186,6 +186,7 @@ def get_identifiers(self):
 # The field name parser is implemented in _string.formatter_field_name_split
 
 class Formatter:
+
```
Comment (dg-pb) [inline Lib/string.py]:
```suggestion
```

File: Lib/test/test_string.py [hunk 7]
```diff
@@ -101,6 +102,22 @@ def test_index_lookup(self):
...
             fmt.format("{0[2]}{0[0]}", {})
 
+    def test_auto_numbering_lookup(self):
+        fmt = string.Formatter()
+        namespace = types.SimpleNamespace(foo=types.SimpleNamespace(bar='baz'))
+        widths = [None, types.SimpleNamespace(qux=4)]
+        self.assertEqual(
+            fmt.format("{.foo.bar:{[1].qux}}", namespace, widths), 'baz ')
+
+    def test_auto_numbering_reenterability(self):
```
Comment (dg-pb) [inline Lib/test/test_string.py:112]: Is this test still needed? There don't seem to be changes to numbering design anymore.

File: Misc/NEWS.d/next/Library/2020-08-07-16-55-57.bpo-27307.Xqzzda.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+Add support for format strings like '{.name}' and '{[1]}' already supported by str.format to string.Formatter
```
Comment (dg-pb) [inline Misc/NEWS.d/next/Library/2020-08-07-16-55-57.bpo-27307.Xqzzda.rst:1]: Add attribute and item access support to :class:`string.Formatter` in auto-numbering mode, which allows format strings like '{.name}' and '{[1]}'.

Comment (dg-pb) [inline Misc/NEWS.d/next/Library/2020-08-07-16-55-57.bpo-27307.Xqzzda.rst]:
```suggestion
Add attribute and item access support to :class:`string.Formatter` in auto-numbering mode, which allows format strings like '{.name}' and '{[1]}'.
```

File: Lib/string.py (hunk 6 above)
Comment (qm2k) [reply Lib/string.py]: Missed it, thank you!

File: Lib/test/test_string.py (hunk 7 above)
Comment (qm2k) [reply Lib/test/test_string.py:112]: We still need reenterability as a feature, and we _could_ have broken it. I will delete it if you insist, but I think this one is good to have.

File: Misc/NEWS.d/next/Library/2020-08-07-16-55-57.bpo-27307.Xqzzda.rst (hunk 8 above)
Comment (qm2k) [reply Misc/NEWS.d/next/Library/2020-08-07-16-55-57.bpo-27307.Xqzzda.rst]: Accepted verbatim.

**Comment by qm2k** [comment]
> > I am starting to remember my motivation for separate class and method
> 
> That was my starting point as well. I wanted to pass `auto_arg_index` to `get_field`. Issue is that it is very hard to do something without breaking backwards compatibility.
> 
> I have a use case in mind for which these changes don't cut it. Namely, `PartialFormatter`, which needs `get_field` to have both original `field_name` and also `auto_arg_index` adjustments.
> 
> Then I thought of something along the lines of `AutoNumber` to pass it easily to methods, so initially it made sense, but eventually realised that this path is a bit tricky.
> 
> Long story short, I will spend some more time, maybe there is some good path forward. If there is not, simply pushing minimal bug fix is probably better than nothing at all.

Have you seen [PEP750](https://peps.python.org/pep-0750/)? I feel like any new work on custom formatters should make use of t-strings (but interestingly some examples there still subclass good old `string.Formatter`).

File: Lib/test/test_string.py (hunk 7 above)
Comment (dg-pb) [reply Lib/test/test_string.py:112]: Ok, after understanding it properly I think the test itself is a good one (even if not directly related). I suggest the below so it is a bit more readable:
```python
class ReenteringFormatter(string.Formatter):
    def format_field(self, value, format_spec):
        if format_spec.isdigit() and int(format_spec) > 0:
            return self.format('{:{}}!', value, int(format_spec) - 1)
        else:
            return super().format_field(value, format_spec)
fmt = ReenteringFormatter()
self.assertEqual(fmt.format('{:{}}', 'X', 3), 'X!!!')
```

Still possible that someone will ask to remove it, but I think it is beneficial.

**Comment by dg-pb** [comment]
> Have you seen [PEP750](https://peps.python.org/pep-0750/)? I feel like any new work on custom formatters should make use of t-strings (but interestingly some examples there still subclass good old `string.Formatter`).

Good point.

t-strings will (hopefully) be great. Unfortunately, it will be a while before I will be able to use them.
t-strings is still in draft - it might take some time before it is out though.
I am not even sure if it is final concept. It had at least 1 radical change in the course of last few (6 maybe) months.
And now I am now on 3.12 and still keep compatibility to 3.8.
Thus, I think `string.Formatter` is still significantly relevant.

In any case, I think this specific PR is useful. From my POV it fixes a bug. See: https://github.com/python/cpython/issues/129273#issuecomment-2613734364

**Review by dg-pb — APPROVED** [review]
LGTM

File: Lib/test/test_string.py (hunk 7 above)
Comment (qm2k) [reply Lib/test/test_string.py:112]: Accepted verbatim.

File: Lib/test/test_string.py [hunk 9]
```diff
@@ -101,6 +102,23 @@ def test_index_lookup(self):
...
+        class ReenteringFormatter(string.Formatter):
+            def format_field(self, value, format_spec):
+                if format_spec.isdigit() and int(format_spec) > 0:
+                    return self.format('{:{}}!', value, int(format_spec) - 1)
+                else:
+                    return super().format_field(value, format_spec)
+        fmt = ReenteringFormatter()
+        self.assertEqual(fmt.format('{:{}}', 'X', 3), 'X!!!')
```
Comment (dg-pb) [inline Lib/test/test_string.py]:
```suggestion
        x = types.SimpleNamespace(a='X')
        print(fmt.format('{.a:{}}', x, 3), 'X!!!')
```

Comment (qm2k) [reply Lib/test/test_string.py]: Good idea, applied.

File: Lib/test/test_string.py [hunk 10]
```diff
@@ -101,6 +102,24 @@ def test_index_lookup(self):
...
+    def test_auto_numbering_reenterability(self):
+        class ReenteringFormatter(string.Formatter):
+            def format_field(self, value, format_spec):
+                if format_spec.isdigit() and int(format_spec) > 0:
+                    return self.format('{:{}}!', value, int(format_spec) - 1)
+                else:
+                    return super().format_field(value, format_spec)
+        fmt = ReenteringFormatter()
+        x = types.SimpleNamespace(a='X')
+        print(fmt.format('{.a:{}}', x, 3), 'X!!!')
```
Comment (encukou) [inline Lib/test/test_string.py]:
```suggestion
        self.assertEqual(fmt.format('{.a:{}}', x, 3), 'X!!!')
```

Comment (qm2k) [reply Lib/test/test_string.py]: Good catch! I need to sleep more...

**Comment by qm2k** [comment]
(Apologies for rebasing in the same change as fixing the issue, [here](https://github.com/python/cpython/compare/1e5587319e85e10e04d17479635d8891a01b0ac9..a493f9cd4ce0e887a7dfc983fc2c14ae1345b796#diff-95bb50b3b847c725c7aa68ed49b82f17a9565d88254dc0afb6b1919ec8662d79) is the relevant change.)

**Comment by encukou** [comment]
No problem; thank you for the contribution!

**Comment by dg-pb** [comment]
@encukou 
I think all of the below can be closed:
* https://github.com/python/cpython/issues/89867
* https://github.com/python/cpython/issues/88662
* https://github.com/python/cpython/issues/71494

**Comment by encukou** [comment]
The reproducer in #89867 still behaves as before. I don't currently have time to look into the details there.

**Comment by encukou** [comment]
I'll treat this as a feature and not backport. Let me know if you want to reconsider.
(Backporting would mean that code tested on 3.13.2 could fail on 3.13.1.)

**Comment by dg-pb** [comment]
> The reproducer in #89867 still behaves as before. I don't currently have time to look into the details there.

You are right, it is not the same thing, I misread it.

However, I had a look at it and that request IMO is a no-go.
I have commented on it with suggestion to close.

The reason for it is clear and unambiguous.
If you have few minutes to spare, I think you could close that one as well.

**Comment by dg-pb** [comment]
> I'll treat this as a feature and not backport.

Fine by me.

Thank you for quick turnaround!
